src/core/thread/csl_tx_scheduler.cpp

```diff
@@ -0,0 +1,273 @@
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "csl_tx_scheduler.hpp"
+
+#if OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE
+
+#include "common/locator-getters.hpp"
+#include "common/logging.hpp"
+#include "common/time.hpp"
+#include "mac/mac.hpp"
+
+namespace ot {
+
+CslTxScheduler::Callbacks::Callbacks(Instance &aInstance)
+    : InstanceLocator(aInstance)
+{
+}
+
+inline otError CslTxScheduler::Callbacks::PrepareFrameForChild(Mac::TxFrame &aFrame,
+                                                               FrameContext &aContext,
+                                                               Child &       aChild)
+{
+    return Get<IndirectSender>().PrepareFrameForChild(aFrame, aContext, aChild);
+}
+
+inline void CslTxScheduler::Callbacks::HandleSentFrameToChild(const Mac::TxFrame &aFrame,
+                                                              const FrameContext &aContext,
+                                                              otError             aError,
+                                                              Child &             aChild)
+{
+    Get<IndirectSender>().HandleSentFrameToChild(aFrame, aContext, aError, aChild);
+}
+
+//---------------------------------------------------------
+//
+CslTxScheduler::CslTxScheduler(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mCslTxChild(nullptr)
+    , mCslTxMessage(nullptr)
+    , mFrameContext()
+    , mCallbacks(aInstance)
+{
+}
+
+void CslTxScheduler::Update(void)
+{
+    if (mCslTxMessage == nullptr)
+    {
+        RescheduleCslTx();
+    }
+    else
+    {
+        // `Mac` has already started the CSL tx, so wait for tx done callback
+        // to call `RescheduleCslTx`
+        if ((mCslTxChild != nullptr) && (mCslTxChild->GetIndirectMessage() != mCslTxMessage))
+        {
+            mCslTxChild                      = nullptr;
+            mCslTxMessage                    = nullptr;
```

Comment:
I don't think we want to set the `mCslTxMessage` to `nullptr` here since it's tracking if we are in middle of CSL tx (i.e., MAC is busy doing CSL tx) and we are still in middle of CSL tx. 

We want `mCslTxMessage` to only be set from `HandleFrameRequest()` and then cleared  from `HandleSentFrame()` so that it tracks the `Mac` operation.